### ISSUE TITLE
Wasmi `v0.32.0-beta.13` seems to have broken linking

### ISSUE THREAD

**Issue by tomaka** [wasmi-labs/wasmi#1018]
I'm trying to update from beta 12 to beta 13 here: https://github.com/smol-dot/smoldot/pull/1817

However, after this PR, wasmi returns linking errors when instantiating something. Example error: `Instantiation("cannot find definition for import env::ext_offchain_index_set_version_1 with type Func(FuncType { params: [I64, I64], results: [] })")`

The way smoldot works is:
 
- It gathers the list of all imported functions using `module.imports()`.
- For each function, [it calls `wasmi::Func::new` and adds the function the linker](https://github.com/smol-dot/smoldot/blob/61cf6a956907f384e8ed3225a4145ff47b8bb26b/lib/src/executor/vm/interpreter.rs#L129-L150).
- Then [it calls `linker.instantiate`](https://github.com/smol-dot/smoldot/blob/61cf6a956907f384e8ed3225a4145ff47b8bb26b/lib/src/executor/vm/interpreter.rs#L177-L178).

I have checked the function signatures and module names and they all look correct. In particular, we indeed add to the linker the functions that wasmi fails to instantiate.

My random guess as what might be causing trouble is the fact that I'm using the generic function constructor: https://github.com/smol-dot/smoldot/blob/61cf6a956907f384e8ed3225a4145ff47b8bb26b/lib/src/executor/vm/interpreter.rs#L132-L135

**Comment by Robbepop** [comment]
@tomaka thanks for reporting the bug, I am going to investigate it with high priority.

> My random guess as what might be causing trouble is the fact that I'm using the generic function constructor:

Is there a specific reason as to why you are not using `Linker::func_new` or even better `Linker::func_wrap`?

**Comment by tomaka** [comment]
> or even better Linker::func_wrap?

I need to be able to accept unresolved imports. If a function name is not recognized by smoldot, the VM should still be able run, and errors should happen only if the function is actually called. To achieve that, I can't know the function signature at compilation time.
And I use the same code for both recognized and unrecognized functions by convenience.

> Does using Linker::func_new or Linker::func_wrap solve your issues?

`func_new` doesn't solve it, no.

**Comment by tomaka** [comment]
> does enabling the new no-hash-maps crate feature of Wasmi solve your issues?

It doesn't compile:

```
error[E0432]: unresolved imports `std::cmp`, `std::mem`, `std::ops`
 --> /Users/tomaka/.cargo/registry/src/index.crates.io-6f17d22bba15001f/wasmi_collections-0.1.0/src/string_interner/detail.rs:4:5
  |
4 |     cmp::Ordering,
  |     ^^^ could not find `cmp` in `std`
5 |     collections::{btree_map::Entry, BTreeMap},
6 |     mem,
  |     ^^^ no `mem` in the root
7 |     ops::Deref,
  |     ^^^ could not find `ops` in `std`
  |
  = help: consider importing one of these items instead:
          core::mem
          std::mem

For more information about this error, try `rustc --explain E0432`.
error: could not compile `wasmi_collections` (lib) due to 1 previous error
```

**Comment by tomaka** [comment]
If I activate both `std` and `no-hash-maps`, the issue is fixed.

**Comment by Robbepop** [comment]
https://github.com/smol-dot/smoldot/pull/1817 udpates from `beta.10` directly to `beta.13`. Can you tell me which was the last working `beta.N`?

**Comment by Robbepop** [comment]
@tomaka 

I took one of the failing smoldot tests and re-created a regression test in the Wasmi repository, however, I somehow cannot reproduce it there. Can you please take a look and maybe tell me what smoldot is doing differently?

```rust
#[test]
fn smoldot_regression() {
    use crate::{Engine, Func, Linker, Memory, MemoryType, Module, Store};
    let wasm = wat::parse_str(
        r#"
        (module
            (import "host" "hello.1" (func $host_hello.1 (param i32) (result i32)))
            (import "env" "memory" (memory $mem 0 4096))
            (func (export "hello") (result i32)
                (call $host_hello.1 (i32.const 3))
                (call $host_hello.2 (i32.const 4))
                (i32.const 2)
                i32.add
            )
        )"#,
    )
    .unwrap();
    let engine = Engine::default();
    let mut linker = <Linker<()>>::new(&engine);
    let mut store = Store::new(&engine, ());
    let memory = Memory::new(&mut store, MemoryType::new(1, Some(4096)).unwrap()).unwrap();
    linker.define("env", "memory", memory).unwrap();
    linker
        .define(
            "host",
            "hello.1",
            Func::new(
                &mut store,
                FuncType::new([ValType::I32], [ValType::I32]),
                |_caller, _params, _results| todo!(),
            ),
        )
        .unwrap();
    let module = Module::new(&engine, &wasm[..]).unwrap();
    linker.instantiate(&mut store, &module).unwrap();
}
```

**Comment by Robbepop** [comment]
I debugged it some more in `smoldot`:

This call fails:
https://github.com/smol-dot/smoldot/blob/61cf6a956907f384e8ed3225a4145ff47b8bb26b/lib/src/executor/vm/interpreter.rs#L178

Because `wasmi::Linker::get("host", "hello")` at this point returns `None` when not enabling `wasmi/no-hash-maps` crate feature. I have not yet figured out why. My base assumption is that this is connected to the `wasmi_collections::StringInterner` which returns consecutive indices for `Symbol` in `no-hash-maps` mode and non-consecutive `Symbol` indices otherwise.

**Comment by Robbepop** [comment]
Even with this test case that mimics `smoldot`'s linker population via `Module::imports` cannot reproduce the issue ...
```rust
#[test]
fn smoldot_regression() {
    use crate::{Engine, Func, Linker, Memory, Module, Store};
    let wasm = wat::parse_str(
        r#"
        (module
            (import "host" "hello" (func $host_hello (param i32) (result i32)))
            (import "env" "memory" (memory $mem 0 4096))
            (func (export "hello") (result i32)
                (call $host_hello (i32.const 3))
                (i32.const 2)
                i32.add
            )
        )"#,
    )
    .unwrap();
    let engine = Engine::default();
    let mut store = Store::new(&engine, ());
    let module = Module::new(&engine, &wasm[..]).unwrap();
    let mut linker = <Linker<()>>::new(&engine);
    for import_type in module.imports() {
        let name = import_type.import_name();
        match import_type.ty() {
            ExternType::Memory(memory_type) => {
                let memory = Memory::new(&mut store, memory_type.clone()).unwrap();
                linker.define(name.module(), name.name(), memory).unwrap();
            }
            ExternType::Func(func_type) => {
                let func = Func::new(
                    &mut store,
                    func_type.clone(),
                    |_caller, _params, _results| unimplemented!(),
                );
                linker.define(name.module(), name.name(), func).unwrap();
            }
            ExternType::Global(_) => unimplemented!(),
            ExternType::Table(_) => unimplemented!(),
        }
    }
    // In smoldot this call fails ...
    linker.instantiate(&mut store, &module).unwrap();
}
```

**Comment by Robbepop** [comment]
@tomaka I created this PR to fix `no_std`+`no-hash-maps` builds which should help you until the underlying issue is properly fixed. https://github.com/wasmi-labs/wasmi/pull/1019

**Comment by Robbepop** [comment]
This minimized Wasmi test case fails only if both `std` and `no-hash-maps` crate features are disabled:
```rust
#[test]
fn min_smoldot_regression() {
    let mut interner = crate::collections::StringInterner::new();
    let sym = interner.get_or_intern("hello");
    assert_eq!(interner.get("hello"), Some(sym));
}
```

**Comment by Robbepop** [comment]
@tomaka I released Wasmi version `0.32.0-beta.14` with the proper fix included.
